src/groupByKey.ts

```diff
@@ -2,6 +2,41 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+type Value = string | number;
+
+type Obj = {
+  [key: string]: Value;
+};
+
+export function groupByKey(items: Obj[], key: keyof Obj): GroupsMap<Obj> {
+  const result: GroupsMap<Obj> = {};
+  const moddedItems = items.map((item1) => {
+    const arrayForItems: Obj[] = [];
+
+    items.filter((item2) => {
+      if (item2[key] === item1[key]) {
+        arrayForItems.push(item2);
+
+        return true;
+      }
+
+      return false;
+    });
+
+    if (arrayForItems.length === 1) {
+      return arrayForItems[0];
+    }
+
+    return arrayForItems;
+  });
+
+  items.forEach((item, index) => {
+    if (Array.isArray(moddedItems[index])) {
+      result[item[key]] = moddedItems[index];
+    } else {
+      result[item[key]] = [moddedItems[index]];
+    }
+  });
```

Comment:
Here, you are iterating over the `items` array again to construct the result object. This is an unnecessary operation since you could construct the result object in a single iteration.